### PR TITLE
Port encryption admin settings to vue

### PR DESCRIPTION
**Pull request by CarlSchwan**
This also remove the button to run a migration since that button was pointing to an URL without a controller and was for ownloud < 8, so probably got removed sometimes ago in the backend but not in the frontend

When enabling

image

When enabled 
image

### CODE REVIEW

File: apps/settings/src/components/Encryption.vue [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+				console.error(errorMessage, error)
+			}
+		},
+	},
+}
+</script>
+
+<style scoped>
+
+.notecard.success {
```
Comment (CarlSchwan) [inline apps/settings/src/components/Encryption.vue:191]: Probably would make sense to move this to a component in nextcloud vue

See discussion in the public design channel :)

@skjnldsv @jancborchardt

File: apps/settings/src/components/Encryption.vue [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	-
+-->
+
+<template>
+	<SettingsSection :title="t('settings', 'Server-side encryption')"
+		:description="t('settings', 'Server-side encryption makes it possible to encrypt files which are uploaded to this server. This comes with limitations like a performance penalty, so enable this only if needed.')"
+		:doc-url="encryptionAdminDoc">
+		<CheckboxRadioSwitch
+			:checked="encryptionEnabled || shouldDisplayWarning"
+			@update:checked="displayWarning()"
```
Comment (Pytal) [inline apps/settings/src/components/Encryption.vue:29]:
```suggestion
			@update:checked="displayWarning"
```

File: apps/settings/src/components/Encryption.vue [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+				<p v-if="encryptionModules.length === 0">
+					{{ t('settings', 'No encryption module loaded, please enable an encryption module in the app menu.') }}
+				</p>
+				<template v-else>
+					<h3>{{ t('settings', 'Select default encryption module:') }}</h3>
+					<fieldset>
+						<CheckboxRadioSwitch v-for="(module, id) in encryptionModules"
+							:checked="module.default"
+							:checked.sync="defaultCheckedModule"
+							@update:checked="checkDefaultModule()"
```
Comment (Pytal) [inline apps/settings/src/components/Encryption.vue:64]:
```suggestion
							@update:checked="checkDefaultModule"
```

File: apps/settings/src/components/Encryption.vue (hunk 2 above)
Comment (skjnldsv) [reply apps/settings/src/components/Encryption.vue:29]: That's a method calling, so `displayWarning()` is correct :)

File: apps/settings/src/components/Encryption.vue [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+				<li>{{ t('settings', 'Be aware that encryption always increases the file size.') }}</li>
+				<li>{{ t('settings', 'It is always good to create regular backups of your data, in case of encryption make sure to backup the encryption keys along with your data.') }}</li>
+			</ul>
+
+			<p class="margin-bottom">
+				{{ t('settings', 'This is the final warning: Do you really want to enable encryption?') }}
+			</p>
+			<Button type="primary"
+				@click=enableEncryption()>
+					{{ t('settings', "Enable encryption") }}
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:49]:
```suggestion
				{{ t('settings', "Enable encryption") }}
```

File: apps/settings/src/components/Encryption.vue [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+			<div v-if="encryptionReady">
+				<p v-if="encryptionModules.length === 0">
+					{{ t('settings', 'No encryption module loaded, please enable an encryption module in the app menu.') }}
+				</p>
+				<template v-else>
+					<h3>{{ t('settings', 'Select default encryption module:') }}</h3>
+					<fieldset>
+						<CheckboxRadioSwitch v-for="(module, id) in encryptionModules"
+							:checked="module.default"
+							:checked.sync="defaultCheckedModule"
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:63]:
```suggestion
							:checked.sync="defaultCheckedModule"
```

Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:63]: Seems odd, most likely sync is not needed here as you're already calling `checkDefaultModule`, so you can change it in th emethod too

File: apps/settings/src/components/Encryption.vue [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+							@update:checked="checkDefaultModule()"
+							:value="id"
+							type="radio"
+							name="default_encryption_module">
+								{{ module.displayName }}
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:68]:
```suggestion
							{{ module.displayName }}
```

File: apps/settings/src/components/Encryption.vue [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+import { loadState } from '@nextcloud/initial-state'
+
+import { generateOcsUrl } from '@nextcloud/router'
+import confirmPassword from '@nextcloud/password-confirmation'
+import { showError } from '@nextcloud/dialogs'
+
+export default {
+	name: 'Encryption',
+	components: {
+		Multiselect,
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:94]:
```suggestion
import CheckboxRadioSwitch from '@nextcloud/vue/dist/Components/CheckboxRadioSwitch'
import Button from '@nextcloud/vue/dist/Components/Button'
import SettingsSection from '@nextcloud/vue/dist/Components/SettingsSection'
import { loadState } from '@nextcloud/initial-state'

import { generateOcsUrl } from '@nextcloud/router'
import confirmPassword from '@nextcloud/password-confirmation'
import { showError } from '@nextcloud/dialogs'

export default {
	name: 'Encryption',
	components: {
```

File: apps/settings/src/components/Encryption.vue [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		CheckboxRadioSwitch,
+		SettingsSection,
+		Button,
+	},
+	data() {
+		const encryptionModules = loadState('settings', 'encryption-modules');
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:100]:
```suggestion
		const encryptionModules = loadState('settings', 'encryption-modules')
```

File: apps/settings/src/components/Encryption.vue [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+			encryptionAdminDoc: loadState('settings', 'encryption-admin-doc'),
+			encryptionModules,
+			shouldDisplayWarning: false,
+			migrating: false,
+			defaultCheckedModule: Object.entries(encryptionModules).find((module) => module[1].default)[0],
+		}
+	},
+	computed: {
+		migrationMessage() {
+			return t('settings', 'You need to migrate your encryption keys from the old encryption (ownCloud <= 8.0) to the new one. Please enable the "Default encryption module" and run {command}', {command: '"occ encryption:migrate"'})
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:115]:
```suggestion
			return t('settings', 'You need to migrate your encryption keys from the old encryption (ownCloud <= 8.0) to the new one. Please enable the "Default encryption module" and run {command}', {
				command: '"occ encryption:migrate"'
			})
```

File: apps/settings/src/components/Encryption.vue [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+				})
+			} catch (e) {
+				this.handleResponse({
+					errorMessage: t('settings', 'Unable to update server side encryption config'),
+					error: e,
+				})
+			}
+		},
+		async checkDefaultModule() {
+			await update('default_encryption_module', this.defaultCheckedModule)
```
Comment (skjnldsv) [inline apps/settings/src/components/Encryption.vue:151]:
```suggestion
			await this.update('default_encryption_module', this.defaultCheckedModule)
```

**Review by skjnldsv — CHANGES_REQUESTED** [review]
See comments

File: apps/settings/src/components/Encryption.vue [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+		<div v-if="encryptionEnabled">
+			<div v-if="encryptionReady">
+				<p v-if="encryptionModules.length === 0">
+					{{ t('settings', 'No encryption module loaded, please enable an encryption module in the app menu.') }}
+				</p>
+				<template v-else>
+					<h3>{{ t('settings', 'Select default encryption module:') }}</h3>
+					<fieldset>
+						<CheckboxRadioSwitch v-for="(module, id) in encryptionModules"
```
Comment (CarlSchwan) [inline apps/settings/src/components/Encryption.vue:61]:
```suggestion
						<CheckboxRadioSwitch v-for="(module, id) in encryptionModules" :key="id"
```

File: apps/settings/src/components/Encryption.vue [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+	- along with this program. If not, see <http://www.gnu.org/licenses/>.
+	-
+-->
+
+<template>
+	<SettingsSection :title="t('settings', 'Server-side encryption')"
+		:description="t('settings', 'Server-side encryption makes it possible to encrypt files which are uploaded to this server. This comes with limitations like a performance penalty, so enable this only if needed.')"
+		:doc-url="encryptionAdminDoc">
+		<CheckboxRadioSwitch
+			:checked="encryptionEnabled || shouldDisplayWarning"
```
Comment (CarlSchwan) [inline apps/settings/src/components/Encryption.vue:28]:
```suggestion
		<CheckboxRadioSwitch :checked="encryptionEnabled || shouldDisplayWarning"
```

File: apps/settings/src/components/Encryption.vue [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+				})
+				this.handleResponse({
+					status: data.ocs?.meta?.status,
+				})
+			} catch (e) {
+				this.handleResponse({
+					errorMessage: t('settings', 'Unable to update server side encryption config'),
+					error: e,
+				})
+			}
```
Comment (CarlSchwan) [inline apps/settings/src/components/Encryption.vue:175]:
```suggestion
```

**Comment by CarlSchwan** [comment]
ping?

File: apps/settings/src/components/Encryption.vue [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+			if (status !== 'ok') {
+				showError(errorMessage)
+				console.error(errorMessage, error)
+			}
+		},
+	},
+}
+</script>
+
+<style scoped>
```
Comment (Pytal) [inline apps/settings/src/components/Encryption.vue:164]:
```suggestion
<style lang="scss" scoped>
```

File: apps/settings/lib/Settings/Admin/Security.php [hunk 15]
```diff
@@ -78,16 +78,32 @@ public function getForm(): TemplateResponse {
...
+
+		$this->initialState->provideInitialState(
+			'encryption-modules',
+			$encryptionModuleList
+		);
+
+		$this->initialState->provideInitialState(
+			'encryption-admin-doc',
+			$this->urlGenerator->linkToDocs('admin-encryption')
+		);
```
Comment (Pytal) [inline apps/settings/lib/Settings/Admin/Security.php:104]: Could we collapse these into 5 lines?

File: apps/settings/src/components/Encryption.vue [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+		<div v-if="encryptionEnabled">
+			<div v-if="encryptionReady">
+				<p v-if="encryptionModules.length === 0">
+					{{ t('settings', 'No encryption module loaded, please enable an encryption module in the app menu.') }}
+				</p>
+				<template v-else>
+					<h3>{{ t('settings', 'Select default encryption module:') }}</h3>
+					<fieldset>
+						<CheckboxRadioSwitch v-for="(module, id) in encryptionModules" :key="id"
```
Comment (artonge) [inline apps/settings/src/components/Encryption.vue:60]: > ':key' should be on a new line

File: apps/settings/src/components/Encryption.vue [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+		const encryptionModules = loadState('settings', 'encryption-modules')
+		return {
+			encryptionReady: loadState('settings', 'encryption-ready'),
+			encryptionEnabled: loadState('settings', 'encryption-enabled'),
+			externalBackendsEnabled: loadState('settings', 'external-backends-enabled'),
+			encryptionAdminDoc: loadState('settings', 'encryption-admin-doc'),
+			encryptionModules,
+			shouldDisplayWarning: false,
+			migrating: false,
+			defaultCheckedModule: Object.entries(encryptionModules).find((module) => module[1].default)[0],
```
Comment (artonge) [inline apps/settings/src/components/Encryption.vue:105]: Move that to a computed property?

File: apps/settings/src/components/Encryption.vue [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+			await this.update('default_encryption_module', this.defaultCheckedModule)
+		},
+		async enableEncryption() {
+			this.encryptionEnabled = true
+			await this.update('encryption_enabled', true)
+		},
+		async handleResponse({ status, errorMessage, error }) {
+			if (status !== 'ok') {
+				showError(errorMessage)
+				console.error(errorMessage, error)
```
Comment (artonge) [inline apps/settings/src/components/Encryption.vue:157]: Remove or use [@nextcloud/logger](https://www.npmjs.com/package/@nextcloud/logger)

**Review by artonge — CHANGES_REQUESTED** [review]

File: apps/settings/src/components/Encryption.vue (hunk 17 above)
Comment (CarlSchwan) [reply apps/settings/src/components/Encryption.vue:112]: this is only computed once, so no need to create a reactive property

**Review by artonge — APPROVED** [review]

**Review by Pytal — APPROVED** [review]

**Pull request merged**